sdks/java/io/pulsar/src/main/java/org/apache/beam/sdk/io/pulsar/ReadFromPulsarDoFn.java

```diff
@@ -0,0 +1,219 @@
+    public double getSize(@Element PulsarSourceDescriptor pulsarSource, @Restriction OffsetRange range) throws PulsarAdminException {
+        //TODO improve getsize estiamate, check pulsar stats to improve get size estimate
+        // https://pulsar.apache.org/docs/en/admin-api-topics/#get-stats
+        double estimateRecords = restrictionTracker(pulsarSource, range).getProgress().getWorkRemaining();
+        return estimateRecords;
+    }
+
+    private Reader<byte[]> newReader(PulsarClient client, String topicPartition) throws PulsarClientException {
+        if(this.readerTst != null) {
+            return this.readerTst;
+        }
+        ReaderBuilder<byte[]> builder = client.newReader().topic(topicPartition).startMessageId(MessageId.earliest);
+        return builder.create();
+    }
+
+    @GetRestrictionCoder
+    public Coder<OffsetRange> getRestrictionCoder() {
+        return new OffsetRange.Coder();
+    }
+
+    @ProcessElement
+    public ProcessContinuation processElement(
+            @Element PulsarSourceDescriptor pulsarSourceDescriptor,
+            RestrictionTracker<OffsetRange, Long> tracker,
+            WatermarkEstimator watermarkEstimator,
+            OutputReceiver<PulsarMessage> output) throws IOException {
+        long startTimestamp = tracker.currentRestriction().getFrom();
+        String topicDescriptor = pulsarSourceDescriptor.getTopic();
+        try (Reader<byte[]> reader = newReader(client, topicDescriptor)) {
+            if (startTimestamp > 0 ) {
+                reader.seek(startTimestamp);
+            }
+            while (true) {
+                if (reader.hasReachedEndOfTopic()) {
+                    reader.close();
+                    return ProcessContinuation.stop();
+                }
+                Message<byte[]> message = reader.readNext();
+                if (message == null) {
+                    return ProcessContinuation.resume();
+                }
+                Long currentTimestamp = message.getPublishTime();
+                // if tracker.tryclaim() return true, sdf must execute work otherwise
+                // doFn must exit processElement() without doing any work associated
+                // or claiming more work
+                //System.out.println(new String(message.getValue(), StandardCharsets.UTF_8));
+                if (!tracker.tryClaim(currentTimestamp)) {
+                    reader.close();
+                    return ProcessContinuation.stop();
+                }
+                if(pulsarSourceDescriptor.getEndMessageId() != null) {
+                    MessageId currentMsgId = message.getMessageId();
+                    boolean hasReachedEndMessageId = currentMsgId.compareTo(pulsarSourceDescriptor.getEndMessageId()) == 0;
+                    if(hasReachedEndMessageId) return ProcessContinuation.stop();
+                }
+                PulsarMessage pulsarMessage = new PulsarMessage(message.getTopicName(), message.getPublishTime(), message);
```

Comment:
I think one of `new PulsarMessage(message.getTopicName(), message.getPublishTime(), message.getData())` (byte[] of payload) or `new PulsarMessage(message.getTopicName(), message.getPublishTime(), message.getValue())` (deserialized payload) would make more sense. 
Or `new PulsarMessage(message)`

Reply:
Please see above comment (same context). I could handle only the message payload for the PulsarMessage, but I decided to get a quick access of the topic and timestamp (which is also the offset being use for each message in beam) and also define the coder for this two (as you can see in PulsarMessageCoder)